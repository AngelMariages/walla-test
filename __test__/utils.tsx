import { render } from '@testing-library/react';
import {
    FavoritesContext,
    FavoritesContextType,
} from 'context/FavoritesContext';
import { FiltersContext, FiltersContextType } from 'context/FiltersContext';

type ContextValues = {
    filtersContext?: Partial<FiltersContextType>;
    favoritesContext?: Partial<FavoritesContextType>;
};

export const renderWithContext = (
    ui: React.ReactElement,
    contextValues?: ContextValues
) => {
    return render(
        <FiltersContext.Provider
            value={{
                search: '',
                setSearch: jest.fn(),
                page: 1,
                incrementPage: jest.fn(),
                sort: {
                    sortBy: 'title',
                    sortOrder: 'asc',
                },
                setSort: jest.fn(),
                ...(contextValues?.filtersContext || {}),
            }}>
            <FavoritesContext.Provider
                value={{
                    favorites: [],
                    favoritesVisible: false,
                    addFavorite: jest.fn(),
                    removeFavorite: jest.fn(),
                    isFavorite: jest.fn(),
                    toggleFavorites: jest.fn(),
                    ...(contextValues?.favoritesContext || {}),
                }}>
                {ui}
            </FavoritesContext.Provider>
        </FiltersContext.Provider>
    );
};
