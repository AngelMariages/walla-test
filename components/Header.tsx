import WallapopLogo from 'public/images/logo-wallapop.svg';
import FavIcon from 'public/icons/fav.svg';
import Search from './SearchBar';
import Link from 'next/link';
import { useContext } from 'react';
import { FiltersContext } from '../context/FiltersContext';

const FavButton = () => {
    return (
        <>
            <button
                className={
                    `fixed bottom-8 right-8 z-10 bg-gray-400  border border-gray-500 shadow-xl shadow-gray-700 rounded-full w-14 h-14 text-center justify-end items-center` +
                    ` md:static md:flex md:flex-row md:h-full md:bg-transparent md:border-none md:shadow-none`
                }>
                <FavIcon className="w-[32px] mx-auto" alt="fav" />
            </button>
        </>
    );
};

const Header = () => {
    const { setSearch, setSort } = useContext(FiltersContext);

    return (
        <>
            <header className="flex sticky top-0 z-10 justify-between items-center w-full h-20 sm:h-24 px-6 sm:px-8 bg-gray-100">
                <div className="flex flex-row justify-start items-center h-full">
                    <Link
                        href="/"
                        className="w-[104px]"
                        onClick={() => setSearch('')}>
                        <WallapopLogo className="w-[104px]" alt="logo" />
                    </Link>
                    <Search />
                </div>
                <FavButton />
            </header>

            <div className="flex flex-row py-4 px-4 justify-center">
                Sort by:
                <div className="flex flex-row pl-4 gap-4 text-white">
                    <button
                        className="px-4 rounded-lg bg-[#12C1AC]"
                        onClick={() => {
                            setSort({
                                sortBy: 'title',
                                sortOrder: 'asc',
                            });
                        }}>
                        Title
                    </button>
                    <button
                        className="px-4 rounded-lg bg-[#12C1AC]"
                        onClick={() => {
                            setSort({
                                sortBy: 'description',
                                sortOrder: 'asc',
                            });
                        }}>
                        Description
                    </button>
                    <button
                        className="px-4 rounded-lg bg-[#12C1AC]"
                        onClick={() => {
                            console.log('price');
                            setSort({
                                sortBy: 'price',
                                sortOrder: 'asc',
                            });
                        }}>
                        Price
                    </button>
                    <button
                        className="px-4 rounded-lg bg-[#12C1AC]"
                        onClick={() => {
                            setSort({
                                sortBy: 'email',
                                sortOrder: 'asc',
                            });
                        }}>
                        Email
                    </button>
                </div>
            </div>
        </>
    );
};

export default Header;
