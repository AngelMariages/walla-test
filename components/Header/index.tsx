import WallapopLogo from 'public/images/logo-wallapop.svg';
import FavIcon from 'public/icons/fav.svg';
import SearchBar from 'components/SearchBar';
import Link from 'next/link';
import { useContext } from 'react';
import { FiltersContext } from 'context/FiltersContext';
import { FavoritesContext } from 'context/FavoritesContext';

const FavButton: React.FC<{
    onClick: () => void;
}> = ({ onClick, ...props }) => {
    return (
        <button
            onClick={onClick}
            className={
                `fixed bottom-8 right-8 z-10 bg-gray-400 rounded-full ring-5 shadow-xl shadow-gray-700 w-14 h-14 text-center justify-end items-center` +
                ` md:static md:flex md:flex-row md:h-full md:bg-transparent md:border-none md:shadow-none`
            }
            {...props}>
            <FavIcon className="w-[32px] fill-red-500 mx-auto" alt="fav" />
        </button>
    );
};

const Header = () => {
    const { setSearch } = useContext(FiltersContext);
    const { toggleFavorites } = useContext(FavoritesContext);

    return (
        <header className="flex sticky top-0 z-10 justify-between items-center w-full h-20 sm:h-24 px-6 sm:px-8 bg-gray-100">
            <div className="flex flex-row justify-start items-center h-full">
                <Link
                    href="/"
                    className="w-[104px]"
                    data-testid="logo"
                    onClick={() => setSearch('')}>
                    <WallapopLogo className="w-[104px]" alt="logo" />
                </Link>
                <SearchBar />
            </div>
            <FavButton data-testid="favbutton" onClick={toggleFavorites} />
        </header>
    );
};

export default Header;
