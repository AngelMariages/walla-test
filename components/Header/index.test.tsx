import { getByTestId } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import { renderWithContext } from '__test__/utils';
import Header from '.';

jest.mock('next/link', () => {
    return function NextLink({
        children,
        ...props
    }: {
        children: React.ReactNode;
    }) {
        return <a {...props}>{children}</a>;
    };
});

jest.mock('next/router', () => ({
    useRouter: () => ({
        push: jest.fn(),
    }),
}));

describe('Header', () => {
    it('should render correctly', () => {
        const { container } = renderWithContext(<Header />);
        expect(container).toMatchSnapshot();
    });

    it('should call toggleFavorites on FavButton click', async () => {
        const toggleFavorites = jest.fn();

        const { container } = renderWithContext(<Header />, {
            favoritesContext: {
                toggleFavorites,
            },
        });

        expect(toggleFavorites).not.toHaveBeenCalled();

        const favButton = getByTestId(container, 'favbutton');
        await act(() => favButton.click());

        expect(toggleFavorites).toHaveBeenCalled();
    });

    it('should call setSearch with empty string on logo click', async () => {
        const setSearch = jest.fn();

        const { container } = renderWithContext(<Header />, {
            filtersContext: {
                setSearch,
            },
        });

        expect(setSearch).not.toHaveBeenCalled();

        const logo = getByTestId(container, 'logo');
        await act(() => logo.click());

        expect(setSearch).toHaveBeenCalledWith('');
    });
});
