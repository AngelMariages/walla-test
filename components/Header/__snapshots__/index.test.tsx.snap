// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<div>
  <header
    class="flex sticky top-0 z-10 justify-between items-center w-full h-20 sm:h-24 px-6 sm:px-8 bg-gray-100"
  >
    <div
      class="flex flex-row justify-start items-center h-full"
    >
      <a
        class="w-[104px]"
        data-testid="logo"
        href="/"
      >
        <svg-url
          alt="logo"
          classname="w-[104px]"
        />
      </a>
      <form
        data-testid="search-form"
      >
        <div
          class="ml-4 md:ml-8 flex gap-2 md:gap-4"
        >
          <input
            class="rounded-md border border-gray-300 px-2 py-1"
            data-testid="search-input"
            placeholder="Search"
            type="text"
          />
          <button
            class="ml-2"
            type="submit"
          >
            <svg-url
              alt="search"
            />
          </button>
        </div>
      </form>
    </div>
    <button
      class="fixed bottom-8 right-8 z-10 bg-gray-400 rounded-full ring-5 shadow-xl shadow-gray-700 w-14 h-14 text-center justify-end items-center md:static md:flex md:flex-row md:h-full md:bg-transparent md:border-none md:shadow-none"
      data-testid="favbutton"
    >
      <svg-url
        alt="fav"
        classname="w-[32px] fill-red-500 mx-auto"
      />
    </button>
  </header>
</div>
`;
